conn_test.go

```diff
@@ -364,33 +364,33 @@ func TestExportKeyingMaterial(t *testing.T) {
 	}
 	c.setLocalEpoch(0)
 
-	_, err := c.ExportKeyingMaterial(exportLabel, nil, 0)
+	_, err := c.state.ExportKeyingMaterial(exportLabel, nil, 0)
```

Comment:
Better to test with `ConnectionState()`

Reply:
Well, that doesn't actually work. I susplect because the `localEpoch` value is invalid and the `serialize` doesn't like that?

I've initially tried this.
```diff
        }
        c.setLocalEpoch(0)
 
-       _, err := c.state.ExportKeyingMaterial(exportLabel, nil, 0)
+       state := c.ConnectionState()
+       _, err := state.ExportKeyingMaterial(exportLabel, nil, 0)
        if err != errHandshakeInProgress {
                t.Errorf("ExportKeyingMaterial when epoch == 0: expected '%s' actual '%s'", errHandshakeInProgress, err)
        }
```

But will get an error:
```
--- FAIL: TestExportKeyingMaterial (0.40s)
panic: interface conversion: interface {} is nil, not uint16 [recovered]
        panic: interface conversion: interface {} is nil, not uint16

goroutine 636 [running]:
testing.tRunner.func1.1(0x69a500, 0xc00009c7e0)
        /home/linuxbrew/.linuxbrew/Cellar/go/1.14/libexec/src/testing/testing.go:941 +0x3d0
testing.tRunner.func1(0xc0000da120)
        /home/linuxbrew/.linuxbrew/Cellar/go/1.14/libexec/src/testing/testing.go:944 +0x3f9
panic(0x69a500, 0xc00009c7e0)
        /home/linuxbrew/.linuxbrew/Cellar/go/1.14/libexec/src/runtime/panic.go:967 +0x15d
github.com/pion/dtls/v2.(*State).serialize(0xc0000ec880, 0xc0000a7968)
        /home/jdebruijn/projects/dtls/state.go:71 +0x2d2
github.com/pion/dtls/v2.(*State).clone(0xc0000ec880, 0x40c808)
        /home/jdebruijn/projects/dtls/state.go:56 +0x43
github.com/pion/dtls/v2.(*Conn).ConnectionState(0xc0000ec840, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/jdebruijn/projects/dtls/conn.go:343 +0x8c
github.com/pion/dtls/v2.TestExportKeyingMaterial(0xc0000da120)
        /home/jdebruijn/projects/dtls/conn_test.go:367 +0x25d
testing.tRunner(0xc0000da120, 0x6faee0)
        /home/linuxbrew/.linuxbrew/Cellar/go/1.14/libexec/src/testing/testing.go:992 +0xdc
created by testing.(*T).Run
        /home/linuxbrew/.linuxbrew/Cellar/go/1.14/libexec/src/testing/testing.go:1043 +0x357
exit status 2
FAIL    github.com/pion/dtls/v2 3.736s
```

Reply:
User can only access `state` via `ConnectionState()`.
It would be better to fix the test data like
```diff
diff --git a/conn_test.go b/conn_test.go
index 8fe3ee2..379e16e 100644
--- a/conn_test.go
+++ b/conn_test.go
@@ -357,14 +357,18 @@ func TestExportKeyingMaterial(t *testing.T) {
 
        c := &Conn{
                state: State{
-                       localRandom:  handshakeRandom{time.Unix(500, 0), rand},
-                       remoteRandom: handshakeRandom{time.Unix(1000, 0), rand},
-                       cipherSuite:  &cipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
+                       localRandom:         handshakeRandom{time.Unix(500, 0), rand},
+                       remoteRandom:        handshakeRandom{time.Unix(1000, 0), rand},
+                       localSequenceNumber: []uint64{0, 0},
+                       cipherSuite:         &cipherSuiteTLSEcdheEcdsaWithAes128GcmSha256{},
                },
        }
        c.setLocalEpoch(0)
+       c.setRemoteEpoch(0)
 
```

Reply:
Fixed now. If there is a better way to do this without having to repeate the `state = c.ConnectionState()` every time, please let me know! I tried replacing it with `c.ConnectionState().ExportKeyingMaterial()` but that isn't valid Go (`ExportKeyingMaterial is not in method set of State`).

```go
	state = c.ConnectionState()
	keyingMaterial, err := state.ExportKeyingMaterial(exportLabel, nil, 10)
```

Reply:
```go
	state = c.ConnectionState()
	keyingMaterial, err := state.ExportKeyingMaterial(exportLabel, nil, 10)
```
seems fine.

BTW, pushed commit seems not updated.

Reply:
You're right. Pushed the correct one now